yoyodyne/models/hard_attention.py

```diff
@@ -0,0 +1,461 @@
+"""Hard monotonic neural HMM classes."""
+
+import argparse
+from typing import Callable, Dict, Optional, Tuple
+
+import numpy
+import torch
+from torch import nn
+
+from .. import data, defaults, util
+from . import lstm, modules
+
+NEG_LOG_EPSILON = -numpy.log(1e-7)
+
+
+class HardAttentionLSTM(lstm.LSTMEncoderDecoder):
+    """Hard Attention Transducer.
+
+    Learns probability distribution of target string by modeling transduction
+    of source string to target string as Markov process. Assumes each character
+    produced is conditioned by state transitions over each source character.
+
+    Default model assumes independence between state and non-monotonic
+    progression over source string. `enforce_monotonic` enforces monotonic
+    state transition (model progresses over each source character), and
+    `attention_context` allows conditioning of state transition over previous
+    _n_ states.
+
+    After:
+        Wu, S. and Cotterell, R. 2019. Exact hard monotonic attention for
+        character-level transduction. In _Proceedings of the 57th Annual
+        Meeting of the Association for Computational Linguistics_, pages
+        1530-1537.
+
+    Original implementation:
+        https://github.com/shijie-wu/neural-transducer
+    """
+
+    enforce_monotonic: bool
+    attention_context: int
+
+    def __init__(
+        self, *args, enforce_monotonic=False, attention_context=0, **kwargs
+    ):
+        """Initializes the encoder-decoder.
+
+        Args:
+            *args: passed to superclass.
+            enforce_monotonic [bool, optional]: Enforce monotonic state
+                transition in decoding.
+            attention_context [int, optional]: Size of context window for
+            conditioning state transition. If 0, state transitions are
+                independent.
+            **kwargs: passed to superclass.
+        """
+        self.enforce_monotonic = enforce_monotonic
+        self.attention_context = attention_context
+        super().__init__(*args, **kwargs)
+        self.classifier = nn.Linear(
+            self.decoder.output_size, self.target_vocab_size
+        )
+        if not self.teacher_forcing:
+            util.log_info(
```

Comment:
I think this should throw an `Error`.

Reply:
I only like throwing Errors if it's going to do behavior that you really shouldn't be doing. If it's something that's just not going to happen I prefer a warning flag.

Reply:
"really shouldn't be doing" -> you're fundamentally changing your architecture or going to have a long term effect that we should end here. That should throw an error since the alternative is building an architecture that's going to break anyhow.

Here, not doing teaching forcing doesn't change architecture or training. There's no long term effects of just changing the flag.